contracts/libraries/Protocol.sol

```diff
@@ -46,10 +63,41 @@ library Protocol {
         uint256 minRequiredMargin;
         uint256 removeLimitOrderFee;
         uint256 minimumOrderNotional;
+        // price cache
+        mapping(uint32 => PriceCache) priceCache;
         // reserved for adding slots in future
         uint256[100] _emptySlots;
     }
 
+    function getBlockNumber() internal view returns (uint32) {
+        return uint32(ArbSys(address(100)).arbBlockNumber());
```

Comment:
This will revert in environments where ArbSys is not present. I see following ways that can be okay:

1. Either we don't use Arbitrum's block number and just use `block.number`.
2. Somehow explore if we can use roundId instead of block.number in the contract?
3. See chainId and only make block number call if it's Arbitrum mainnet or testnet, else fallback to block.number.
4. Screw everything and store cache in oracle make the getPrice call on it as non-view (which will involve upstream calls to be non-view as well). This sounds bad but it may work.

What do you think can work best? I think 1) would be most clean and 2) the most accurate. Does using l2 block number make a difference? I believe even virtual price in Uniswap would use block.number hence it would be off as well. You can see arbitrum as a chain with 2 min block time (or whatever their posting interval is). Tx is not confirmed when it lands in the inbox, it's confirmed when l1 tx is done, technically sequencer can reorg the inbox list as they want which can be like setting the l2 block number at their will. I think we should aim for the decentralised version of Arbitrum.

For 2) we might have to have `updatePoolPriceCache` on the chain link oracle, and maybe expose two price methods: use cached price and use latest price. At beginning of every call we can ask the oracle to be updated.